tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Grpc/GrpcLegacy/Client/AsyncCallHandleUnaryResponseInstrumentation.cs

```diff
@@ -0,0 +1,102 @@
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Grpc.GrpcLegacy.Client
+{
+    /// <summary>
+    /// Grpc.Core.Internal calltarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "Grpc.Core",
+        TypeName = "Grpc.Core.Internal.AsyncCall`2",
+        MethodName = "HandleUnaryResponse",
+        ReturnTypeName = ClrNames.Void,
+        ParameterTypeNames = new[] { ClrNames.Bool, "Grpc.Core.Internal.ClientSideStatus", "Grpc.Core.Internal.IBufferReader", "Grpc.Core.Metadata" },
+        MinimumVersion = "2.0.0",
+        MaximumVersion = "2.*.*",
+        IntegrationName = nameof(Grpc))]
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static class AsyncCallHandleUnaryResponseInstrumentation
+    {
+        internal static CallTargetState OnMethodBegin<TTarget, TStatus, TBufferReader, TMetadata>(
+            TTarget instance,
+            bool success,
+            in TStatus clientSideStatus,
+            in TBufferReader receivedMessageReader,
+            in TMetadata responseHeaders)
+        {
+            // receivedStatus can change in the middle of this method, so we need to wait for the method to finish,
+            // to grab the final status.
+            var tracer = Tracer.Instance;
+            if (GrpcCoreApiVersionHelper.IsSupported && tracer.Settings.IsIntegrationEnabled(IntegrationId.Grpc))
+            {
+                // using CreateFrom to avoid boxing ClientSideStatus struct
+                var receivedStatus = DuckType.CreateCache<ClientSideStatusWithMetadataStruct>.CreateFrom(clientSideStatus);
+                var status = receivedStatus.Status;
+                var asyncCall = instance.DuckCast<AsyncCallStruct>();
+                var scope = GrpcLegacyClientCommon.CreateClientSpan(tracer, in asyncCall.Details, in status);
+                if (scope?.Span is { } span)
+                {
+                    if (receivedStatus.Trailers is { Count: > 0 })
+                    {
+                        span.SetHeaderTags(new MetadataHeadersCollection(receivedStatus.Trailers), tracer.Settings.GrpcTags, defaultTagPrefix: GrpcCommon.ResponseMetadataTagPrefix);
+                    }
+                    else if (responseHeaders is not null)
+                    {
+                        var responseMetadata = responseHeaders.DuckCast<IMetadata>();
+                        if (responseMetadata.Count > 0)
+                        {
+                            span.SetHeaderTags(new MetadataHeadersCollection(responseMetadata), tracer.Settings.GrpcTags, defaultTagPrefix: GrpcCommon.ResponseMetadataTagPrefix);
+                        }
+                    }
+                }
+
+                // need to pass the finish time so that you don't get traces
+                // where the child span closes after the parent
+                return new CallTargetState(scope, state: DateTimeOffset.UtcNow);
```

Comment:
`CallTargetState` has a separate field/constructor that you can use to store the `DateTimeOffset` value without boxing:

```suggestion
                return new CallTargetState(scope, state: null, startTime: DateTimeOffset.UtcNow);
```

Reply:
😮 TIL, awesome, thanks! Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/2545/commits/e64d79e17f9a80e7221741aba276d09227d30107